### PR TITLE
Modifications in rcnn.cpp for more applicable outputs of instance segmentation. 

### PR DESCRIPTION
**Pull request by HaiyangPeng**
When using the rcnn.cpp to do inference, i find that the outputs of instance segmentation are on a padded input image rather than an original image. This will be inconvenient for me or other users to directly apply the results in specific cases. Therefore, I modified this cpp which supports the following functions now:
-  this cpp is consistent with the detectron2 training code where padding is not used. Only needs to set the vars of INPUT_H_ and INPUT_W_ to be same as the original training data. (e.g., if use cityscapes, INPUT_H_=1024, INPUT_W_=2048). The ResizeShortestEdge op will be applied automatically.
- additionally, this code also supports when the input image size is not same as the training data. For this, padding is applied. Of course, the instance segmentation results will be corrected according to the padding and resizing ops.
-  main codes are not modified.
This code has been tested on cityscapes, validating the effectiveness on different input image sizes. Any questions are appreciated ^-^.

### CODE REVIEW

**Comment by wang-xinyu** [comment]
@freedenS can you please review this?

**Comment by wang-xinyu** [comment]
@freedenS kindly please check this PR

**Comment by freedenS** [comment]
> @freedenS kindly please check this PR

ok, i'll take a look

File: rcnn/rcnn.cpp [hunk 1]
```diff
@@ -18,10 +18,16 @@ static const std::vector<float> PIXEL_STD = {1.0, 1.0, 1.0};
...
 static constexpr int NUM_CLASSES = 80;
-static constexpr int INPUT_H = 480;
-static constexpr int INPUT_W = 640;
-static int IMAGE_HEIGHT = 800;
-static int IMAGE_WIDTH = 1333;
+static int INPUT_H;  // size of model input
+static int INPUT_W;
+static constexpr int INPUT_H_ = 480;  // size of original image
+static constexpr int INPUT_W_ = 640;
+static int IMAGE_HEIGHT;  // size of model input
```
Comment (freedenS) [inline rcnn/rcnn.cpp:25]: looks like IMAGE_HEIGHT are same with INPUT_H, one left will be better if so

File: rcnn/rcnn.cpp [hunk 2]
```diff
@@ -63,40 +69,39 @@ const std::vector<float>& aspect_ratios) {
...
-    width = static_cast<int>(neww + 0.5);
-    auto resize_layer = network->addResize(input);
-    assert(resize_layer);
-    resize_layer->setResizeMode(ResizeMode::kLINEAR);
-    resize_layer->setOutputDimensions(Dims3{ height, width, c });
-    IMAGE_HEIGHT = height;
-    IMAGE_WIDTH = width;
+    cv::Mat re(h, w, CV_8UC3);
+    cv::resize(img, re, re.size(), 0, 0, cv::INTER_LINEAR);
+    cv::Mat out(input_h, input_w, CV_8UC3, cv::Scalar(128, 128, 128));
```
Comment (freedenS) [inline rcnn/rcnn.cpp:89]: looks like you do the center-padding with 128, but rcnn-c4 dont need padding and rcnn-fpn use lower-right padding with 0 (i remember)

File: rcnn/rcnn.cpp [hunk 3]
```diff
@@ -63,40 +69,39 @@ const std::vector<float>& aspect_ratios) {
...
-        newh = ratio * height;
-        neww = MIN_SIZE;
-    }
-    if (std::max(newh, neww) > MAX_SIZE) {
-        ratio = MAX_SIZE / static_cast<float>(std::max(newh, neww));
-        newh = newh * ratio;
-        neww = neww * ratio;
+        w = r_h * img.cols;
+        h = input_h;
+        x = (input_w - w) / 2;
```
Comment (freedenS) [inline rcnn/rcnn.cpp:84]: is it ok if (input_w - w) is an odd number

**Comment by HaiyangPeng** [comment]
Thanks for reviewing this PR, and I'll reply all your questions and suggestions later. BTW, there are so many months, so I need to review my codes ^-^

File: rcnn/rcnn.cpp (hunk 1 above)
Comment (HaiyangPeng) [reply rcnn/rcnn.cpp:25]: Yes, these two vars are same. I'll leave only one var.

File: rcnn/rcnn.cpp (hunk 2 above)
Comment (HaiyangPeng) [reply rcnn/rcnn.cpp:89]: The padding operation is imported from the function preprocessImg() in your previous common.hpp file, so you actually have used padding in rcnn-c4.

**Comment by HaiyangPeng** [comment]
I have modified the codes according to the suggestions of @freedenS. Do I need to create a new PR for this? @wang-xinyu

File: rcnn/rcnn.cpp [hunk 4]
```diff
@@ -510,18 +524,29 @@ int main(int argc, char** argv) {
         auto end = std::chrono::system_clock::now();
         std::cout << std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << "ms" << std::endl;
 
-        float h_ratio = static_cast<float>(INPUT_H) / IMAGE_HEIGHT;
-        float w_ratio = static_cast<float>(INPUT_W) / IMAGE_WIDTH;
+        float h_ratio = static_cast<float>(h_ori) / (INPUT_H - Y_PAD * 2);  // ratio of original image size to model input size
```
Comment (freedenS) [inline rcnn/rcnn.cpp:527]: is't ok if batch!=1, maybe h_ori will be different with each other batch

File: rcnn/rcnn.cpp (hunk 2 above)
Comment (freedenS) [reply rcnn/rcnn.cpp:89]: in fact it's not really preprocessing, its purpose is just make sure the input be same size, so that no matter what is the size of original image, they will be same when they are feed to the model. The really preprocess is ITensor* DataPreprocess(INetworkDefinition *network, ITensor& input) which is inserted into trt engine.
so, the ratio of the postprocess will be equal

File: rcnn/rcnn.cpp (hunk 4 above)
Comment (HaiyangPeng) [reply rcnn/rcnn.cpp:527]: Yes. If batch!=1, each h_ori may be different. So, I will fix this var.

File: rcnn/rcnn.cpp (hunk 2 above)
Comment (HaiyangPeng) [reply rcnn/rcnn.cpp:89]: hah, the main purpose is simple (outputting detections and segmentations on the original images, not the padded images).

File: rcnn/rcnn.cpp (hunk 4 above)
Comment (HaiyangPeng) [reply rcnn/rcnn.cpp:527]: Yes, I will modify it according to your above thinking of padding.

**Comment by freedenS** [comment]
what's the use of INPUT_H_ and INPUT_W_? looks like they limit the max size of input

**Comment by HaiyangPeng** [comment]
> what's the use of INPUT_H_ and INPUT_W_? looks like they limit the max size of input

The original size of images, e.g., in cityscapes, INPUT_H_ = 1024, INPUT_W_ = 2048

**Comment by freedenS** [comment]
BTW, can you keep the calibrate.hpp pace with the new preprocess?

**Comment by HaiyangPeng** [comment]
> BTW, can you keep the calibrate.hpp pace with the new preprocess?

The only used function in calibrate.hpp is Int8EntropyCalibrator2. It seems that Int8EntropyCalibrator2 is used to do configure int8? Can you simply introduce the function of it?

**Comment by freedenS** [comment]
> > BTW, can you keep the calibrate.hpp pace with the new preprocess?
> 
> The only used function in calibrate.hpp is Int8EntropyCalibrator2. It seems that Int8EntropyCalibrator2 is used to do configure int8? Can you simply introduce the function of it? Oh, I find you use preprocessImg() in Int8EntropyCalibrator2. So, just change it?

just do the same preprocess (before feeded to model) in function {bool Int8EntropyCalibrator2::getBatch(void* bindings[], const char* names[], int nbBindings)}, so the quantization will be compatible with inference

File: rcnn/common.hpp [hunk 5]
```diff
@@ -77,23 +77,32 @@ static inline int read_files_in_dir(const char *p_dir_name, std::vector<std::str
...
+    int w, h;
+    float x, y;
     float r_w = input_w / (img.cols*1.0);
     float r_h = input_h / (img.rows*1.0);
     if (r_h > r_w) {
         w = input_w;
         h = r_w * img.rows;
-        x = 0;
+        x = 0.0;
         y = (input_h - h) / 2;
```
Comment (freedenS) [inline rcnn/common.hpp:88]: the problem still exists though y is float, because input_h, h are all int.
one way is y = (input_h - h) / 2.f;

Comment (freedenS) [reply rcnn/common.hpp:88]: looks like this function haven't be called

Comment (HaiyangPeng) [reply rcnn/common.hpp:88]: No, this function will be called by {bool Int8EntropyCalibrator2::getBatch(void* bindings[], const char* names[], int nbBindings)} as mentioned by you. You can check it.

**Comment by freedenS** [comment]
could you try to merge preprocessImg and preprocessImg_, because they are so similar.
one way i think is to let X_LEFT_PAD be  referenced parameter.
and could you remove the print for debug?
that's all, thank you very much😊

**Comment by HaiyangPeng** [comment]
> could you try to merge preprocessImg and preprocessImg_, because they are so similar. one way i think is to let X_LEFT_PAD be referenced parameter. and could you remove the print for debug? that's all, thank you very much😊

OK，I will merge them to one function in common.hpp as you previously wrote and delete all the debug codes. THX too. Have a good holiday ^-^

**Comment by HaiyangPeng** [comment]
I have modified the codes according to all suggestions of @freedenS , please check them again, THX. @wang-xinyu

**Comment by freedenS** [comment]
> > could you try to merge preprocessImg and preprocessImg_, because they are so similar. one way i think is to let X_LEFT_PAD be referenced parameter. and could you remove the print for debug? that's all, thank you very much😊
> 
> OK，I will merge them to one function in common.hpp as you previously wrote and delete all the debug codes. THX too. Have a good holiday ^-^

thank you! you too!

**Comment by freedenS** [comment]
X_LEFT_PAD these four var need update in preprocessImg
one way is static inline cv::Mat preprocessImg(cv::Mat& img, int input_w, int input_h, int& X_LEFT_PAD, ...)

**Comment by HaiyangPeng** [comment]
> 



> X_LEFT_PAD these four vars need update in preprocessImg one way is static inline cv::Mat preprocessImg(cv::Mat& img, int input_w, int input_h, int& X_LEFT_PAD, ...)

Oh, I have checked it. Definitely, I need the reference passing to make the four vars update. THX, I have modified it, please check.

**Comment by freedenS** [comment]
last question, can you please check the tutorial in readme? and do you need to add something new in it?
that's all, thank you!

**Comment by HaiyangPeng** [comment]
> last question, can you please check the tutorial in readme? and do you need to add something new in it? that's all, thank you!

OK, I have updated the README 😊.

**Comment by freedenS** [comment]
@wang-xinyu i have no advice for this commit now.

**Comment by wang-xinyu** [comment]
Thanks guys. merging it.
